Support ssh agent

justone opened the pull request:
Thanks for building such a useful tool.

One thing that was missing was the ability to use SSH agent authorization instead of private keys.

This is my attempt to add that functionality.  Basically, if there is no key on an instance it will try to use the SSH_AUTH_SOCK variable to communicate with an agent.  If there is a key on an instance, it will use that key as before.  And, finally, if the `--agent` flag is passed, the agent will be used without looking for a key.

Please let me know what you think.

simcap commented on ssh/ssh.go:
```diff
@@ -169,20 +173,39 @@ func DecryptSSHKey(key []byte, password []byte) (gossh.Signer, error) {
-		var passphrase []byte
-		passphrase, err = terminal.ReadPassword(int(syscall.Stdin))
+func resolveAuthMethod(priv privateKey) (gossh.AuthMethod, error) {
+	var authmethod gossh.AuthMethod
```
Currently this method will return a `nil` authmethod. 

Here is the fix (I tested it):

```go
	if len(priv.body) > 0 {
		signer, err := gossh.ParsePrivateKey(priv.body)
		if err != nil && strings.Contains(err.Error(), "cannot decode encrypted private keys") {
			fmt.Fprintf(os.Stderr, "This SSH key is encrypted. Please enter passphrase for key '%s':", priv.path)
			var passphrase []byte
			passphrase, err = terminal.ReadPassword(int(syscall.Stdin))
			if err != nil {
				return nil, err
			}
			fmt.Fprintln(os.Stderr)
			signer, err = DecryptSSHKey(priv.body, passphrase)
		}

		return gossh.PublicKeys(signer), err
	}

	sshAuthSock := os.Getenv("SSH_AUTH_SOCK")
	if len(sshAuthSock) == 0 {
		return nil, fmt.Errorf("No key provided and no SSH_AUTH_SOCK env variable set, unable to resolve auth")
	}

	agentUnixSock, err := net.Dial("unix", sshAuthSock)
	if err != nil {
		return nil, err
	}

	return gossh.PublicKeysCallback(agent.NewClient(agentUnixSock).Signers), nil
```

simcap commented on ssh/ssh.go:
```diff
@@ -169,20 +173,39 @@ func DecryptSSHKey(key []byte, password []byte) (gossh.Signer, error) {
+			authmethod = gossh.PublicKeys(signer)
+		}
+
+		return authmethod, err
```
Basically here can return `nil` `authmethod`. See fix above.

simcap commented:
@justone Thanks for the work. We were actually about to add `ssh-agent` support as it is really useful.

I started the review. Couple of issues that will need to be addressed. See inline comments

justone commented:
I updated the `resolveAuthMethod` function, and reworked the error handling on the key decryption part to handle non-encrypted-key errors and any errors from `DecryptSSHKey`.

simcap commented:
Thanks for the update of the `resolveAuthMethod`. There is just one fix pending and it should be good to merge after a final review. Thanks.

justone commented:
I took care of the second fix (the suggestion to return nil) by flipping the conditional at the top of that block.  Now, the method only continues if a password is required.  Any other errors are just returned:

```
if !strings.Contains(err.Error(), "cannot decode encrypted private keys") {
    return nil, err
}
```

Other errors are handled the same way, so at the end the `signer` is valid and can be returned.

simcap commented:
@justone Yes `resolveAuthMethod` was fine. I was not clear though.

 I was talking about this https://github.com/wallix/awless/pull/106/files#diff-3ca3b34b111bd186138d20eb2bc8d8adR220

The effect is testing locally with one of my instance: `awless ssh broker_1 --agent` this works;  `awless ssh broker_1` this does not work.

I might merge you PR in a feature branch where we can commit both of us maybe and I could do this fix.

justone commented:
Ah, ok.  That conditional is there so that the code doesn't try to find a key that has no name.

Can you share the line for the `broker_1` instance from `awless show instances`? I'm wondering if there is an ssh key there too that is getting in the way when the `--agent` flag is missing.

Perhaps it would be better to rework the code so that multiple authmethods are created and tried in turn.  So if there was a key specified and the ssh-agent is available it will try the key first and then fall back to the ssh-agent.  And if there is only a key specified, then that's all that is tried.

I believe you should be able to push to my branch, I have the "Allow edits from maintainers" option checked.  Let me know if that works.

simcap commented:
Yes. As said in my previous comment, and as you gathered, there is a ssh key getting in the way. More precisely, what is getting in the way is just the key name of the AWS keypair (ssh key) associated with the instance. 

Also as you point out a good implementation would be a chain of fallback on the AuthMethod. Trying with the given key (or not) and then on last resort contact the ssh-agent.

I'll let you see if you want to do the implementation, then I can join later on to do/finish or review it on this branch as you have allow edit from maintainers.

Anyway thanks for tackling this issue!

Here if you want to see the `awless show broker_1`:

|    PROPERTY      |                              VALUE                               |
|-------------------|------------------------------------------------------------------|
| Architecture      | x86_64                                                           |
| Hypervisor        | xen                                                              |
| ID                | i-0dc3954d1b9e13bde                                              |
| Image             | ami-36230256                                                     |
| KeyPair           | simonus                                                          |
| Name              | broker_1                                                         |
....
| Uptime            | 6 days                                                           |
| Vpc               | vpc-da4297be                                                     |
| Zone              | us-west-1a                                                       |

#### Relations:
us-west-1[region]
	↳ @awless-demo-vpc[vpc]
		↳ @awless-demo-subnet[subnet]
			↳ @broker_1[instance]

Depending on: simonus[keypair], @KafkaSecurityGroup[securitygroup], @SSHSecurityGroup[securitygroup], vol-0174a0a96895470d5[volume]

Siblings: @broker_2[instance], @broker_3[instance], @zookeeper[instance]

justone commented:
Cool, thanks for the confirmation.  I'll take a swing at the fallback strategy tomorrow.

justone commented:
@simcap I attempted an implementation of multiple authmethods.  Please let me know what you think.

Thanks.

justone commented:
@simcap any thoughts?

simcap commented:
Thanks. It works for the usual cases:

- `awless ssh broker_1` (no ssh agent running, but key present)
- `awless ssh broker_1` (key only loaded in running agent) 
- `awless ssh --print-cli`
- `awless ssh --print-config`
- `awless ssh machine_i_have_no_keys_aws_for`
- `awless ssh machine_with_no_aws_keys`

I have some remarks though:
- we need to provide some logging info context to the user now that we use multiple methods of ssh auth . For example, basically print which methods was successfully used and/or which fail withouth being to verbose.
- also the way it is implemented it is no easy to know programmatically where we are at and it the current ongoing auth (or the auth just done) was through the ssh-agent. (Just as an example, it could be used to remove the last line `IdentityFile` when using the `--print-config`)

Anyway we can merge this PR (unless you feel you want to implement the remarks above, but I can do it later on the master)

justone commented:
Cool. I understand both points. I'll work on it tomorrow morning.

justone commented:
Sorry to take so long, but I got a chance to work on the code and got things working.  Please take a look at let me know what you think.

Here is some sample output.  I have my ssh agent running, but the key matches.

```
$ ls TestAwless*
TestAwless  TestAwless.pub

$ awless list instances
|        ID ▲         |    ZONE    |       NAME       |  STATE  |   TYPE    |   PUBLIC IP   |  PRIVATE IP   |  UPTIME   |  KEYPAIR   |
|---------------------|------------|------------------|---------|-----------|---------------|---------------|-----------|------------|
| i-044831c02f3f3bf84 | us-west-2c | Tester           | running | t2.micro  | 54.69.105.120 | 172.31.13.55  | 49 mins   | TestAwless |

$ awless ssh Tester --print-config
[info]    trying with username(s) [ec2-user, ubuntu, centos, bitnami, admin, root] and auth method(s) [keypair TestAwless, ssh agent]

Host Tester
        Hostname 54.69.105.120
        User ec2-user
        IdentityFile TestAwless

$ awless ssh Tester --print-cli
[info]    trying with username(s) [ec2-user, ubuntu, centos, bitnami, admin, root] and auth method(s) [keypair TestAwless, ssh agent]
/usr/bin/ssh ec2-user@54.69.105.120 -i TestAwless

$ awless ssh Tester
[info]    trying with username(s) [ec2-user, ubuntu, centos, bitnami, admin, root] and auth method(s) [keypair TestAwless, ssh agent]
[info]    Login as 'ec2-user' on '54.69.105.120', using keypair TestAwless for auth with ssh client '/usr/bin/ssh'

Last login: Sat Jun 24 00:37:52 2017 from ec2-52-32-167-187.us-west-2.compute.amazonaws.com

       __|  __|_  )
       _|  (     /   Amazon Linux AMI
      ___|\___|___|

https://aws.amazon.com/amazon-linux-ami/2017.03-release-notes/
4 package(s) needed for security, out of 6 available
Run "sudo yum update" to apply all updates.
[ec2-user@ip-172-31-13-55 ~]$ 
```

simcap commented:
Thanks. A few things to iron out in term of getting more compact logging messages and fix display errors in verbose logging.

But I will merge that and fix it later on.

Cheers

justone commented:
Cool, thank you!
